Non-flat Numpy arrays throw error on compression

A while back I noticed that trying to compress non-flat Numpy arrays no longer works.
```
______________________________________________________________________ test_round_trip_decompress_into[u8-900] ______________________________________________________________________

length = 900, dtype = 'u8'

    @pytest.mark.parametrize("length", all_lengths)
    @pytest.mark.parametrize("dtype", all_dtypes)
    def test_round_trip_decompress_into(length, dtype):
        data = np.random.uniform(0, 1000, size=length).astype(dtype).reshape(-1, 1)
>       compressed = standalone.simple_compress(data, ChunkConfig())
E       TypeError: 'ndarray' object cannot be converted to 'PyArray<T, D>'

pco_python/test/test_standalone.py:20: TypeError
============================================================================== short test summary info ==============================================================================
FAILED pco_python/test/test_standalone.py::test_round_trip_decompress_into[u8-900] - TypeError: 'ndarray' object cannot be converted to 'PyArray<T, D>'
=========================================================================== 18 failed, 39 passed in 0.93s ===========================================================================
```
I don't know if this change was intentional or not. If so, the documentation for the Python function should be updated as it says that "any shape" is allowed:
``` rust
pub fn register(m: &Bound<PyModule>) -> PyResult<()> {
  /// Compresses an array into a standalone format.
  ///
  /// :param nums: numpy array to compress. This may have any shape.
  ///   However, it must be contiguous, and only the following data types are
  ///   supported: float16, float32, float64, int16, int32, int64, uint16, uint32, uint64.
  /// :param config: a ChunkConfig object containing compression level and
  ///   other settings.

--- a/pco_python/src/standalone.rs
+++ b/pco_python/src/standalone.rs
@@ -67,7 +67,7 @@ pub fn register(m: &Bound<PyModule>) -> PyResult<()> {
     match_number_enum!(
       number_type,
       NumberType<T> => {
-        simple_compress_generic(py, nums.downcast::<PyArray1<T>>()?, &config)
+        simple_compress_generic(py, utils::downcast_to_flat::<T>(nums)?, &config)
       }
     )
   }
@@ -96,7 +96,7 @@ pub fn register(m: &Bound<PyModule>) -> PyResult<()> {
     match_number_enum!(
       number_type,
       NumberType<T> => {
-        simple_decompress_into_generic(py, compressed, dst.downcast::<PyArray1<T>>()?)
+        simple_decompress_into_generic(py, compressed, utils::downcast_to_flat::<T>(dst)?)
       }
     )
   }

--- a/pco_python/src/standalone.rs
+++ b/pco_python/src/standalone.rs
@@ -47,9 +47,9 @@ fn simple_decompress_into_generic<T: Number + Element>(
 pub fn register(m: &Bound<PyModule>) -> PyResult<()> {
   /// Compresses an array into a standalone format.
   ///
-  /// :param nums: numpy array to compress. This may have any shape.
-  ///   However, it must be contiguous, and only the following data types are
-  ///   supported: float16, float32, float64, int16, int32, int64, uint16, uint32, uint64.
+  /// :param nums: numpy array to compress. This must be 1D, contiguous, and
+  ///   only the following data types are supported: float16, float32, float64,
+  ///   int16, int32, int64, uint16, uint32, uint64.
   /// :param config: a ChunkConfig object containing compression level and
   ///   other settings.
   ///
@@ -76,8 +76,8 @@ pub fn register(m: &Bound<PyModule>) -> PyResult<()> {
   /// Decompresses pcodec compressed bytes into a pre-existing array.
   ///
   /// :param compressed: a bytes object a full standalone file of compressed data.
-  /// :param dst: a numpy array to fill with the decompressed values. May have
-  ///   any shape, but must be contiguous.
+  /// :param dst: a numpy array to fill with the decompressed values. Must be
+  ///   both 1D and contiguous.
   ///
   /// :returns: progress, an object with a count of elements written and
   ///   whether the compressed data was finished. If dst is shorter than the

--- a/pco_python/src/utils.rs
+++ b/pco_python/src/utils.rs
@@ -1,8 +1,27 @@
-use numpy::{PyArrayDescr, PyArrayDescrMethods};
-use pco::data_types::NumberType;
+use std::any;
+
+use numpy::{
+  Element, PyArray1, PyArrayDescr, PyArrayDescrMethods, PyUntypedArray, PyUntypedArrayMethods,
+};
+use pco::data_types::{Number, NumberType};
 use pco::errors::PcoError;
 use pyo3::exceptions::{PyRuntimeError, PyTypeError};
+use pyo3::prelude::*;
 use pyo3::{Bound, PyErr, PyResult, Python};
+
+pub fn downcast_to_flat<'a, T: Number + Element>(
+  arr: &'a Bound<PyUntypedArray>,
+) -> PyResult<&'a Bound<'a, PyArray1<T>>> {
+  let res = arr.downcast::<PyArray1<T>>().map_err(|_e| {
+    PyTypeError::new_err(format!(
+      "{}D {} numpy array could not be cast to 1D {} array",
+      arr.shape().len(),
+      arr.dtype().to_string(),
+      any::type_name::<T>(),
+    ))
+  })?;
+  Ok(res)
+}
 
 pub fn core_dtype_from_str(s: &str) -> PyResult<NumberType> {
   match s.to_uppercase().as_str() {

--- a/pco_python/src/wrapped/compressor.rs
+++ b/pco_python/src/wrapped/compressor.rs
@@ -62,9 +62,9 @@ impl PyFc {
   ///
   /// This does the bulk of the work of compression.
   ///
-  /// :param nums: numpy array to compress. This may have any shape.
-  ///   However, it must be contiguous, and only the following data types are
-  ///   supported: float16, float32, float64, int16, int32, int64, uint16, uint32, uint64.
+  /// :param nums: numpy array to compress. This must be 1D, contiguous, and
+  ///   only the following data types are supported: float16, float32, float64,
+  ///   int16, int32, int64, uint16, uint32, uint64.
   /// :param config: a ChunkConfig object containing compression level and
   ///   other settings.
   ///

--- a/pco_python/src/wrapped/compressor.rs
+++ b/pco_python/src/wrapped/compressor.rs
@@ -74,7 +74,7 @@ impl PyFc {
   fn chunk_compressor(
     &self,
     py: Python,
-    nums: Bound<PyUntypedArray>,
+    nums: &Bound<PyUntypedArray>,
     config: &PyChunkConfig,
   ) -> PyResult<PyCc> {
     let config = config.try_into()?;
@@ -82,7 +82,7 @@ impl PyFc {
     match_number_enum!(
       number_type,
       NumberType<T> => {
-        let cc = self.chunk_compressor_generic::<T>(py, nums.downcast::<PyArray1<T>>()?, &config)?;
+        let cc = self.chunk_compressor_generic::<T>(py, utils::downcast_to_flat::<T>(&nums)?, &config)?;
         Ok(PyCc(cc))
       }
     )

--- a/pco_python/test/test_standalone.py
+++ b/pco_python/test/test_standalone.py
@@ -173,12 +173,12 @@ def test_decompress_without_n_hint():
 
 
 def test_multidimensional():
-    nums = np.random.normal(size=[10, 11])
+    nums = np.random.normal(size=(10, 11))
     with pytest.raises(TypeError, match="cannot be converted"):
         standalone.simple_compress(nums, ChunkConfig())
 
 
 def test_non_contiguous():
-    nums = np.random.normal(size=[20])[::2]
+    nums = np.random.normal(size=20)[::2]
     with pytest.raises(TypeError, match="not contiguous"):
         standalone.simple_compress(nums, ChunkConfig())

--- a/pco_python/test/test_standalone.py
+++ b/pco_python/test/test_standalone.py
@@ -170,3 +170,15 @@ def test_decompress_without_n_hint():
         compressed = f.read()
 
     assert len(standalone.simple_decompress(compressed)) == 2000
+
+
+def test_multidimensional():
+    nums = np.random.normal(size=[10, 11])
+    with pytest.raises(TypeError, match="cannot be converted"):
+        standalone.simple_compress(nums, ChunkConfig())
+
+
+def test_non_contiguous():
+    nums = np.random.normal(size=[20])[::2]
+    with pytest.raises(TypeError, match="not contiguous"):
+        standalone.simple_compress(nums, ChunkConfig())

--- a/pco_python/test/test_standalone.py
+++ b/pco_python/test/test_standalone.py
@@ -173,8 +173,10 @@ def test_decompress_without_n_hint():
 
 
 def test_multidimensional():
-    nums = np.random.normal(size=(10, 11))
-    with pytest.raises(TypeError, match="cannot be converted"):
+    nums = np.random.normal(size=[10, 11])
+    with pytest.raises(
+        TypeError, match="2D float64 numpy array could not be cast to 1D"
+    ):
         standalone.simple_compress(nums, ChunkConfig())
 
 

